handler/src/main/java/io/netty/handler/ssl/SniHandler.java

```diff
@@ -59,127 +64,161 @@ public SniHandler(DomainNameMapping<? extends SslContext> mapping) {
+                                    //        case false:
+                                    //            struct {};
+                                    //        case true:
+                                    //            Extension extensions<0..2^16-1>;
+                                    //    };
+                                    // } ClientHello;
+                                    //
+                                    int offset = readerIndex + 43;
+
+                                    int sessionIdLength = in.getUnsignedByte(offset);
+                                    offset += sessionIdLength + 1;
+
+                                    int cipherSuitesLength = in.getUnsignedShort(offset);
+                                    offset += cipherSuitesLength + 2;
+
+                                    int compressionMethodLength = in.getUnsignedByte(offset);
+                                    offset += compressionMethodLength + 1;
+
+                                    int extensionsLength = in.getUnsignedShort(offset);
+                                    offset += 2;
+                                    int extensionsLimit = offset + extensionsLength;
+
+                                    while (offset < extensionsLimit) {
+                                        int extensionType = in.getUnsignedShort(offset);
+                                        offset += 2;
+
+                                        int extensionLength = in.getUnsignedShort(offset);
+                                        offset += 2;
+
+                                        // SNI
+                                        // See https://tools.ietf.org/html/rfc6066#page-6
+                                        if (extensionType == 0) {
+                                            int serverNameType = in.getUnsignedByte(offset + 2);
+                                            if (serverNameType == 0) {
+                                                int serverNameLength = in.getUnsignedShort(offset + 3);
+                                                String hostname = in.toString(offset + 5, serverNameLength,
+                                                        CharsetUtil.UTF_8);
+                                                select(ctx, IDN.toASCII(hostname,
+                                                        IDN.ALLOW_UNASSIGNED).toLowerCase(Locale.US));
+                                                return;
+                                            } else {
+                                                // invalid enum value
+                                                break loop;
+                                            }
+                                        }
+
+                                        offset += extensionLength;
+                                    }
+                                    break loop;
+                                } else {
```

Comment:
I think if you structure this the opposite way, you can get rid of the `break loop` above:

```java
if (in.readableBytes() < packetLength) {
    // client hello incomplete try again to decode once more data is ready.
    return;
}

...
// Fall-through
```